refactor: port tensorflow model to pytorch

aledelunap opened the pull request:
closes #8 

The pytorch port includes the following:
- dataset.py -> Dataset implementation with pytorch lightning 
- model.py -> UNet implementation with pytorch lightning. The LWBNA UNet implementation is pending, the current pull request includes a submodule for the UNet architecture
- training.py -> Training stage with pytorch lightning, model is saved as *.pth

aledelunap committed:
add submodule for unet

aledelunap committed:
build: UNet model

aledelunap committed:
build: dataset and dataloader in pytorch

aledelunap committed:
docs: author name correction

aledelunap committed:
build: random point selection for an image

aledelunap committed:
build: UNet model

aledelunap committed:
build: training stage

aledelunap committed:
refactor: UNet now implemented in pytorch

aledelunap committed:
refactor: project now implemented in pytorch

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+        x, y = get_example(self.data_dir, fiber_file)
+
+        x = torch.tensor(x)
+        y = torch.tensor(y)[None]
```
I am ignorant of why the None is used. What does it mean?

aledelunap replied:
It just adds a dimension to end up with the shape (1, n, n) to prepare for batching

aledelunap replied:
This is already answered in another comment.

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+    idx = 0
+
+    for file in os.listdir(data_dir):
+        if file[: len(stage)] == stage and file[-6:] == "params":
```
Consider using .startswith(stage) and .endswith("params") that are easier to read.

aledelunap replied:
perfect!

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+from PIL import Image, ImageOps
+
+
+def get_fibers_dictionary(data_dir: str, stage: str):
```
Do I understand correctly that it returns a dictionary mapping the images of a data split? Why is it needed to use a dictionary if the keys are integers?

aledelunap replied:
the integers are needed in order to use the dataloader directly, so the dictionary keeps track of the mapping between the idx (integer) and the respective image

aledelunap replied:
I will delete this function, I now see there is no need for a dictionary

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class FiberDataset(torch.utils.data.Dataset):
+    def __init__(self, stage: str, fiber_dic: dict, data_dir: str):
```
How is fiber_dict supposed to look like? Maybe add it in the docstring

aledelunap replied:
fiber_dict will be removed

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+        x, y = get_example(self.data_dir, fiber_file)
+
+        x = torch.tensor(x)
+        y = torch.tensor(y)[None]
```
Why is `[None]` needed?

aledelunap replied:
to add an extra dimension for batching (1, n, n), I will refactor it to just move the axis of the original image instead of squeezing it

fcossio commented on src/dataset.py:
```diff
@@ -0,0 +1,134 @@
+                self.data_dir,
+            )
+
+    def train_dataloader(self: pl.LightningDataModule):
```
I don't think self needs to be typed, as it can be deduced from the class, but maybe I'm wrong?

aledelunap replied:
the course types it as such (I don't know if it a standard thing), I can remove it

fcossio commented on src/training.py:
```diff
@@ -0,0 +1,76 @@
+from pytorch_lightning import Trainer, seed_everything
+
+from model import UNet
+from dataset import FiberDataModule
```
Maybe this can go inside the fiberdiameter folder so we can import it as fiberdiameter.dataset and it becomes easier to test in the future

fcossio commented on src/training.py:
```diff
@@ -0,0 +1,76 @@
+    return args
+
+
+def init_from_valid_args(cls, args):
```
Curiosity question: what is the use of this module?

aledelunap replied:
it initializes the class cls with the respective arguments contained in args since the parser receives as input variables for different classes

fcossio commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "src/unet_pytorch"]
+	path = src/unet_pytorch
+	url = git@github.com:jaxony/unet-pytorch.git
```
Not v important: Is it possible to specify the hash of the commit hash that is being used?

fcossio commented on pyproject.toml:
```diff
@@ -2,7 +2,12 @@
 version = "0.1.1"
 description = "Measure the diameter of fibers 10x faster!"
-authors = ["Fernando Cossio", "Yael Suarez", "Subhadeep Chowdhury", "Luis Cossio", "Ale de Luna"]
+authors = [
```
please add the required dependencies to the package with

`poetry add <X>` # if it will be used by the package
`poetry add --group dev <X>` # if it will be used for dev only

aledelunap replied:
done

fcossio reviewed (changes requested):
Great work! I have left some comments in the files.

Please don't delete the files that are being deleted and put the model and dataloaders inside the dataset and segmentation folders.

aledelunap committed:
build: restored deleted files

aledelunap committed:
refactor: remove use of fiber_dict

- The number of images for each stage is taken as input from the user
- The file structure for each stage is assumed to be {stage}.{idx},
where idx is a 4 integer number with leading 0s
- Now the total number of images is used to generate the idx and create
the image path

aledelunap committed:
refactor: remove fiber_dict, add number of workers
of machine

aledelunap committed:
refactor: move files to segmentation folder and
rename model.py file to UNet.py

aledelunap committed:
refactor: mode file to dataset folder

aledelunap committed:
refactor: move submodule to segmentation folder

aledelunap committed:
refactor: moved file to segmentatin folder

aledelunap committed:
refactor: change addition of axis  to y as [None]
to the get_example function

aledelunap committed:
build: addition of torch port dependencies

fcossio reviewed (approved):
Nice! great first PR
